File: src/engine/vk/swapchain.c
#include "swapchain.h"
#include "device.h"
#include "image.h"
#include "physical_device.h"
#include "vk.h"
#include "engine/utils/math.h"

#include <stb_ds.h>

static VkSurfaceCapabilitiesKHR
_getSurfaceCapabilitiesKHR(VkPhysicalDevice phys, VkSurfaceKHR surface)
{
    VkSurfaceCapabilitiesKHR capabilities;
    vkCheck(vkGetPhysicalDeviceSurfaceCapabilitiesKHR(phys, surface, &capabilities))
    {
        ecs_abort(1, "Failed to get physical devce surface capabilities");
    }
    return capabilities;
}

static int _calcNumImages(VkSurfaceCapabilitiesKHR capabilities, int requestedImages)
{
    int maxImages = capabilities.maxImageCount;
    int minImages = capabilities.minImageCount;
    int result = minImages;
    if (maxImages != 0) {
        result = i32min(requestedImages, maxImages);
    }
    result = i32max(result, minImages);
    ecs_trace("Requested [%d] images, got [%d] images", requestedImages, result);
    ecs_trace("Surface capabilities, maxImages: [%d], minImages: [%d]", maxImages, minImages);
    return result;
}

static SurfaceFormat
_calcSurfaceFormat(VkPhysicalDevice phys, VkSurfaceKHR surface)
{
    uint32_t count;
    vkCheck(vkGetPhysicalDeviceSurfaceFormatsKHR(phys, surface, &count, NULL))
    {
        ecs_abort(1, "Failed to get count of surface formats");
    }
    if (count == 0) {
        ecs_abort(1, "No surface format retrieved");
    }
    VkSurfaceFormatKHR formats[count];
    vkCheck(vkGetPhysicalDeviceSurfaceFormatsKHR(phys, surface, &count, formats))
    {
        ecs_abort(1, "Failed to get surface formats");
    }
    SurfaceFormat format = { .imageFormat = VK_FORMAT_B8G8R8A8_SRGB,
        .colorSpace = formats[0].colorSpace };
    for (uint32_t i = 0; i < count; ++i) {
        VkSurfaceFormatKHR f = formats[i];
        if (f.format == VK_FORMAT_B8G8R8A8_SRGB
            && f.colorSpace == VK_COLOR_SPACE_SRGB_NONLINEAR_KHR) {
            format.imageFormat = f.format;
            format.colorSpace = f.colorSpace;
            break;
        }
    }
    return format;
}

static VkExtent2D
_calcSwapchainExtent(VkSurfaceCapabilitiesKHR capabilities, uint32_t defaultWidth, uint32_t defaultHeight)
{
    VkExtent2D extent;
    if (capabilities.currentExtent.width == 0xFFFFFFFF) {
        // If current extent is not defined, use default
        extent.width = u32clamp(defaultWidth, capabilities.minImageExtent.width, capabilities.maxImageExtent.width);
        extent.height = u32clamp(defaultHeight, capabilities.minImageExtent.height, capabilities.maxImageExtent.height);
    } else {
        // Otherwise use current extent
        extent = capabilities.currentExtent;
    }
    return extent;
}

typedef struct {
    int aspectMask;
    int baseArrayLayer;
    int format;
    int layerCount;
    int mipLevels;
    int viewType;
} ImageViewData;

static ImageViewData
_defaultImageViewCreateInfo()
{
    return (ImageViewData) {
        .baseArrayLayer = 0,
        .layerCount = 1,
        .mipLevels = 1,
        .viewType = VK_IMAGE_VIEW_TYPE_2D,
    };
}

static VkImageView
_VkImageView(VkDevice device, VkImage image, ImageViewData data)
{
    VkImageViewCreateInfo ci = {
        .sType = VK_STRUCTURE_TYPE_IMAGE_VIEW_CREATE_INFO,
        .image = image,
        .viewType = data.viewType,
        .format = data.format,
        .subresourceRange = (VkImageSubresourceRange) {
            .aspectMask = data.aspectMask,
            .baseMipLevel = 0,
            .levelCount = data.mipLevels,
            .baseArrayLayer = data.baseArrayLayer,
            .layerCount = data.layerCount,
        },
    };
    VkImageView view;
    vkCheck(vkCreateImageView(device, &ci, NULL, &view))
    {
        ecs_abort(1, "Failed to create image view");
    }
    return view;
}

static ImageView*
_newImageViews(VkDevice device, VkSwapchainKHR swapchain, int format)
{
    uint32_t count;
    VkImage* images = NULL;
    ImageView* views = NULL;

    vkCheck(vkGetSwapchainImagesKHR(device, swapchain, &count, NULL))
    {
        ecs_abort(1, "Failed to get number of surface images");
    }
    arrsetlen(images, count);
    arrsetlen(views, count);
    vkCheck(vkGetSwapchainImagesKHR(device, swapchain, &count, images))
    {
        ecs_abort(1, "Failed to get surface images");
    }
    ImageViewData data = _defaultImageViewCreateInfo();
    data.format = format;
    data.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT;
    for (uint32_t i = 0; i < count; ++i) {
        VkFenceCreateInfo ci = {
            .sType = VK_STRUCTURE_TYPE_FENCE_CREATE_INFO,
            .flags = VK_FENCE_CREATE_SIGNALED_BIT
        };
        VkFence fence;
        vkCheck(vkCreateFence(device, &ci, NULL, &fence))
        {
            ecs_abort(1, "Failed to create fence");
        }
        VkSemaphoreCreateInfo semaphoreCI = {
            .sType = VK_STRUCTURE_TYPE_SEMAPHORE_CREATE_INFO,
        };
        VkSemaphore imgAcquisitionSemaphore, renderCompleteSemaphore;
        vkCheck(vkCreateSemaphore(device, &semaphoreCI, NULL, &imgAcquisitionSemaphore))
        {
            ecs_abort(1, "Failed to create image acquisition semaphore");
        }
        vkCheck(vkCreateSemaphore(device, &semaphoreCI, NULL, &renderCompleteSemaphore))
        {
            ecs_abort(1, "Failed to create render complete semaphore");
        }
        VkImageView imgView = _VkImageView(device, images[i], data);
        views[i]
            = (ImageView) {
                  .handle = imgView,
                  .image = images[i],
                  .fence = fence,
                  .acquisitionSemaphore = imgAcquisitionSemaphore,
                  .renderCompleteSemaphore = renderCompleteSemaphore
              };
        ecs_trace("Created VkImageView = %#p on VkSwapchainKHR = %#p", imgView, swapchain);
    }
    return views;
}

void createSwapchain(const SwapchainCreateInfo* pCreateInfo, Swapchain* pSwapchain)
{
    const Device* renderDevice = pCreateInfo->pDevice;
    const PhysicalDevice* physDev = renderDevice->phys;
    VkPhysicalDevice vkPhysDev = physDev->vkPhysicalDevice;
    VkDevice device = renderDevice->vkDevice;
    ecs_trace("Creating Swapchain on device [%#p]", device);
    ecs_log_push();

    VkSurfaceCapabilitiesKHR capabilities = _getSurfaceCapabilitiesKHR(vkPhysDev, pCreateInfo->surface);
    int numImages = _calcNumImages(capabilities, pCreateInfo->requestedImages);
    SurfaceFormat format = _calcSurfaceFormat(vkPhysDev, pCreateInfo->surface);
    VkExtent2D extent = _calcSwapchainExtent(capabilities, pCreateInfo->defaultWidth, pCreateInfo->defaultHeight);

    VkSwapchainCreateInfoKHR ci = {
        .sType = VK_STRUCTURE_TYPE_SWAPCHAIN_CREATE_INFO_KHR,
        .surface = pCreateInfo->surface,
        .minImageCount = numImages,
        .imageFormat = format.imageFormat,
        .imageColorSpace = format.colorSpace,
        .imageExtent = extent,
        .imageArrayLayers = 1,
        .imageUsage = VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT,
        .imageSharingMode = VK_SHARING_MODE_EXCLUSIVE,
        .preTransform = capabilities.currentTransform,
        .compositeAlpha = VK_COMPOSITE_ALPHA_OPAQUE_BIT_KHR,
        .clipped = true,
    };
    if (pCreateInfo->vSync) {
        ci.presentMode = VK_PRESENT_MODE_FIFO_KHR;
    } else {
        ci.presentMode = VK_PRESENT_MODE_IMMEDIATE_KHR;
    }
    VkSwapchainKHR swapchain;
    vkCheck(vkCreateSwapchainKHR(device, &ci, NULL, &swapchain))
    {
        ecs_abort(1, "Failed to create swapchain");
    }
    ecs_trace("VkSwapchainKHR = %#p", swapchain);

    ImageView* views = _newImageViews(device, swapchain, format.imageFormat);

    ecs_log_pop();
    *pSwapchain = (Swapchain) {
        .handle = swapchain,
        .device = renderDevice,
        .surface = pCreateInfo->surface,
        .arrViews = views,
        .format = format,
        .currentFrame = 0,
        .extent = extent,
    };
}

const ImageView* swapchainCurrentView(const Swapchain* swapchain)
{
    return &swapchain->arrViews[swapchain->currentFrame];
}

bool swapchainAcquire(Swapchain* swapchain)
{
    bool resize = false;
    const ImageView* view = swapchainCurrentView(swapchain);
    switch (vkAcquireNextImageKHR(swapchain->device->vkDevice, swapchain->handle, -0L,
        view->acquisitionSemaphore, NULL, &swapchain->currentFrame)) {
    case VK_ERROR_OUT_OF_DATE_KHR:
        resize = true;
        break;
    case VK_SUBOPTIMAL_KHR:
        resize = true;
        break;
    case VK_SUCCESS:
        break;
    default:
        ecs_abort(1, "Failed to acquire image")
    }
    return resize;
}

bool swapchainPresent(Swapchain* swapchain, VkQueue queue)
{
    bool resize = false;

    const ImageView* view = swapchainCurrentView(swapchain);
    ecs_trace("Presenting to image [%d]", swapchain->currentFrame);
    VkPresentInfoKHR presentInfo
        = {
              .sType = VK_STRUCTURE_TYPE_PRESENT_INFO_KHR,
              .pWaitSemaphores = &view->renderCompleteSemaphore,
              .waitSemaphoreCount = 1,
              .swapchainCount = 1,
              .pSwapchains = &swapchain->handle,
              .pImageIndices = &swapchain->currentFrame,
          };
    switch (vkQueuePresentKHR(queue, &presentInfo)) {
    case VK_ERROR_OUT_OF_DATE_KHR:
        resize = true;
        break;
    case VK_SUBOPTIMAL_KHR:
        break;
    case VK_SUCCESS:
        break;
    default:
        ecs_abort(1, "Failed to present KHR");
    }
    swapchain->currentFrame = (swapchain->currentFrame + 1) % arrlen(swapchain->arrViews);
    return resize;
}

File: src/main/kotlin/gfx/vk/Swapchain.kt
package gfx.vk

import gfx.Window
import gfx.vk.VulkanUtils.Companion.vkCheck
import org.lwjgl.system.MemoryStack
import org.lwjgl.system.MemoryUtil
import org.lwjgl.vulkan.*
import org.lwjgl.vulkan.VK10.VK_SUCCESS
import org.tinylog.kotlin.Logger
import java.nio.LongBuffer
import kotlin.math.max
import kotlin.math.min

class Swapchain(device: Device, surface: Surface, window: Window, requestedImages: Int, vsync: Boolean) {
    val device: Device
    val frameInfo: Array<FrameInfo>
    val surfaceFormat: SurfaceFormat
    val swapChainExtent: VkExtent2D
    val vkSwapChain: Long
    var currentFrame: Int
    val numImages: Int get() = frameInfo.size

    init {
        Logger.debug("Creating Vulkan swapchain")
        this.device = device
        MemoryStack.stackPush().use { stack ->
            val physicalDevice = device.physicalDevice
            // Get surface capabilities
            val surfCapabilities = VkSurfaceCapabilitiesKHR.calloc(stack)
            vkCheck(
                KHRSurface.vkGetPhysicalDeviceSurfaceCapabilitiesKHR(
                    device.physicalDevice.vkPhysicalDevice,
                    surface.vkSurface, surfCapabilities
                ), "Failed to get surface capabilities"
            )
            val numImages = calcNumImages(surfCapabilities, requestedImages)
            surfaceFormat = calcSurfaceFormat(physicalDevice, surface)
            swapChainExtent = calcSwapChainExtent(window, surfCapabilities)
            val vkSwapchainCreateInfo = VkSwapchainCreateInfoKHR.calloc(stack)
                .`sType$Default`()
                .surface(surface.vkSurface)
                .minImageCount(numImages)
                .imageFormat(surfaceFormat.imageFormat)
                .imageColorSpace(surfaceFormat.colorSpace)
                .imageExtent(swapChainExtent)
                .imageArrayLayers(1)
                .imageUsage(VK13.VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT)
                .imageSharingMode(VK13.VK_SHARING_MODE_EXCLUSIVE)
                .preTransform(surfCapabilities.currentTransform())
                .compositeAlpha(KHRSurface.VK_COMPOSITE_ALPHA_OPAQUE_BIT_KHR)
                .clipped(true)
            if (vsync) {
                vkSwapchainCreateInfo.presentMode(KHRSurface.VK_PRESENT_MODE_FIFO_KHR)
            } else {
                vkSwapchainCreateInfo.presentMode(KHRSurface.VK_PRESENT_MODE_IMMEDIATE_KHR)
            }
            val lp = stack.mallocLong(1)
            vkCheck(
                KHRSwapchain.vkCreateSwapchainKHR(device.vkDevice, vkSwapchainCreateInfo, null, lp),
                "Failed to create swap chain"
            )
            vkSwapChain = lp[0]

            frameInfo = createFrameInfos(stack, device, vkSwapChain, surfaceFormat.imageFormat)
            currentFrame = 0
        }
    }

    fun cleanup() {
        Logger.debug("Destroying Vulkan swapchain")
        frameInfo.forEach { it.cleanup() }
        KHRSwapchain.vkDestroySwapchainKHR(device.vkDevice, vkSwapChain, null)
    }

    fun acquireNextImage(): Boolean {
        var resize = false
        MemoryStack.stackPush().use { stack ->
            val ip = stack.mallocInt(1)
            val err = KHRSwapchain.vkAcquireNextImageKHR(
                device.vkDevice, vkSwapChain, 0.inv(),
                frameInfo[currentFrame].syncSemaphores.imgAcquisitionSemaphore.vkSemaphore, MemoryUtil.NULL, ip
            )
            if (err == KHRSwapchain.VK_ERROR_OUT_OF_DATE_KHR) {
                resize = true
            } else if (err == KHRSwapchain.VK_SUBOPTIMAL_KHR) {
                resize = true
            } else if (err != VK_SUCCESS) {
                throw RuntimeException("Failed to acquire image: $err")
            }
            currentFrame = ip[0]
        }
        return resize
    }

    fun presentImage(vkQueue: VkQueue): Boolean {
        var resize = false
        MemoryStack.stackPush().use { stack ->
            val present = VkPresentInfoKHR.calloc(stack)
                .`sType$Default`()
                .pWaitSemaphores(
                    stack.longs(
                        frameInfo[currentFrame].syncSemaphores.renderCompleteSemaphore.vkSemaphore
                    )
                )
                .swapchainCount(1)
                .pSwapchains(stack.longs(vkSwapChain))
                .pImageIndices(stack.ints(currentFrame))
            val err = KHRSwapchain.vkQueuePresentKHR(vkQueue, present)
            if (err == KHRSwapchain.VK_ERROR_OUT_OF_DATE_KHR) {
                resize = true
            } else if (err == KHRSwapchain.VK_SUBOPTIMAL_KHR) {
                resize = true
            } else if (err != VK13.VK_SUCCESS) {
                throw RuntimeException("Failed to present KHR: $err")
            }
        }
        currentFrame = (currentFrame + 1) % numImages
        return resize
    }

    private fun createFrameInfos(stack: MemoryStack, device: Device, swapChain: Long, format: Int): Array<FrameInfo> {
        val swapChainImages = getSwapchainImages(stack, device, swapChain)
        val numImages = swapChainImages.capacity()
        val imageViewData = ImageView.Info().format(format).aspectMask(VK13.VK_IMAGE_ASPECT_COLOR_BIT)
        return Array(numImages) {
            FrameInfo(
                swapChainImages[it],
                ImageView.create(device, swapChainImages[it], imageViewData),
                SyncSemaphores.create(device)
            )
        }
    }

    private fun getSwapchainImages(
        stack: MemoryStack,
        device: Device,
        swapChain: Long
    ): LongBuffer {
        val ip = stack.mallocInt(1)
        vkCheck(
            KHRSwapchain.vkGetSwapchainImagesKHR(device.vkDevice, swapChain, ip, null),
            "Failed to get number of surface images"
        )
        val numImages = ip[0]
        val swapChainImages = stack.mallocLong(numImages)
        vkCheck(
            KHRSwapchain.vkGetSwapchainImagesKHR(device.vkDevice, swapChain, ip, swapChainImages),
            "Failed to get surface images"
        )
        return swapChainImages
    }


    companion object {
        private fun calcNumImages(surfCapabilities: VkSurfaceCapabilitiesKHR, requestedImages: Int): Int {
            val maxImages = surfCapabilities.maxImageCount()
            val minImages = surfCapabilities.minImageCount()
            var result = minImages
            if (maxImages != 0) {
                result = min(requestedImages, maxImages)
            }
            result = max(result, minImages)
            Logger.debug(
                "Requested [{}] images, got [{}] images. Surface capabilities, maxImages: [{}], minImages: [{}]",
                requestedImages, result, maxImages, minImages
            )
            return result
        }

        private fun calcSurfaceFormat(physicalDevice: PhysicalDevice, surface: Surface): SurfaceFormat {
            var imageFormat: Int
            var colorSpace: Int
            MemoryStack.stackPush().use { stack ->
                val ip = stack.mallocInt(1)
                vkCheck(
                    KHRSurface.vkGetPhysicalDeviceSurfaceFormatsKHR(
                        physicalDevice.vkPhysicalDevice,
                        surface.vkSurface,
                        ip,
                        null
                    ),
                    "Failed to get number of surface formats"
                )
                val numFormats = ip[0]
                if (numFormats <= 0) {
                    throw RuntimeException("No surface format retrieved")
                }
                val surfaceFormats = VkSurfaceFormatKHR.calloc(numFormats, stack)
                vkCheck(
                    KHRSurface.vkGetPhysicalDeviceSurfaceFormatsKHR(
                        physicalDevice.vkPhysicalDevice,
                        surface.vkSurface,
                        ip,
                        surfaceFormats
                    ),
                    "Failed to get surface formats"
                )
                imageFormat = VK13.VK_FORMAT_B8G8R8A8_SRGB
                colorSpace = surfaceFormats[0].colorSpace()
                for (i in 0 until numFormats) {
                    val surfaceFormatKHR = surfaceFormats[i]
                    if (surfaceFormatKHR.format() == VK13.VK_FORMAT_B8G8R8A8_SRGB &&
                        surfaceFormatKHR.colorSpace() == KHRSurface.VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
                    ) {
                        imageFormat = surfaceFormatKHR.format()
                        colorSpace = surfaceFormatKHR.colorSpace()
                        break
                    }
                }
            }
            return SurfaceFormat(imageFormat, colorSpace)
        }

        fun calcSwapChainExtent(
            window: Window,
            surfCapabilities: VkSurfaceCapabilitiesKHR
        ): VkExtent2D {
            val result = VkExtent2D.calloc()
            if (surfCapabilities.currentExtent().width().toLong() == 0xFFFFFFFF) {
                // Surface size undefined. Set to the window size if within bounds
                var width = min(window.width, surfCapabilities.maxImageExtent().width())
                width = max(width, surfCapabilities.minImageExtent().width())
                var height = min(window.height, surfCapabilities.maxImageExtent().height())
                height = max(height, surfCapabilities.minImageExtent().height())
                result.width(width)
                result.height(height)
            } else {
                // Surface already defined, just use that for the swap chain
                result.set(surfCapabilities.currentExtent())
            }
            return result
        }
    }

    data class SurfaceFormat(val imageFormat: Int, val colorSpace: Int)

    data class SyncSemaphores(
        val imgAcquisitionSemaphore: Semaphore,
        val renderCompleteSemaphore: Semaphore
    ) {
        fun cleanup() {
            imgAcquisitionSemaphore.cleanup()
            renderCompleteSemaphore.cleanup()
        }

        companion object {
            fun create(device: Device): SyncSemaphores {
                return SyncSemaphores(Semaphore.create(device), Semaphore.create(device, ))
            }
        }
    }

    data class FrameInfo(val image: Long, val view: ImageView, val syncSemaphores: SyncSemaphores) {
        fun cleanup() {
            view.cleanup()
            syncSemaphores.cleanup()
        }
    }
}